xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -434,24 +488,76 @@ absl::Status ProcessWindowedEinsumLoopForActivationCaching(
           dus_boundary_constant->shape(), slice_indices));
       VLOG(5) << "Created slice op for second slice: "
               << slice_indices->ToString();
-      // The slice we need this time is the output of the first
-      // collective-permute
-      HloInstruction* cp_output;
-      for (HloInstruction* gte_user : input_gte->users()) {
-        if (gte_user->opcode() == HloOpcode::kCollectivePermute) {
-          cp_output = gte_user;
-          break;
-        }
-      }
       new_full_buffer_output =
           while_body->AddInstruction(HloInstruction::CreateDynamicUpdateSlice(
               full_buffer_output_gte->shape(), full_buffer_output_gte,
-              cp_output,
+              first_cp_output,
               {dus_boundary_constant, slice_indices, dus_boundary_constant}));
     }
+
+    // If we have a DS(parameter_index1), then operands are sharded along the
+    // contracting dim. Slice indices will be the contracting dim's slices.
+    HloInstruction* slice_index;
```

Comment:
You are matching a very specific pattern here. Are you sure this is the right thing to do to fix this bug or does the issue occur for other patterns, too?

Reply:
Yea it's a good point. I simplified it a bit and made it more generic. We still need to be specific when we look for the slice index when we update the final output buffer when the loop is unrolled. I added more comment to explain why we need to be specific in that case.